lib/Logger/LogContext.h

```diff
@@ -0,0 +1,559 @@
+  Derived const& self() const noexcept {
+    return static_cast<Derived const&>(*this);
+  }
+};
+
+template <class Overloads>
+struct LogContext::OverloadVisitor : GenericVisitor<OverloadVisitor<Overloads>>,
+                                     Overloads {
+  explicit OverloadVisitor(Overloads overloads)
+      : GenericVisitor<OverloadVisitor>(), Overloads(std::move(overloads)) {}
+  template <class T>
+  void visit(std::string_view const& key, T&& value) const {
+    this->operator()(key, std::forward<T>(value));
+  }
+};
+
+template <class Vals, const char... Keys[]>
+struct LogContext::ValueBuilder {
+  template <const char Key[], class Value>
+  auto with(Value v) && {
+    auto vals =
+        std::tuple_cat(std::move(_vals), std::make_tuple(std::forward<Value>(v)));
+    return ValueBuilder<decltype(vals), Keys..., Key>(std::move(vals));
+  }
+  std::shared_ptr<Values> share() && {
+    return std::make_shared<ValuesImpl<Vals, Keys...>>(std::move(_vals));
+  }
+
+ private:
+  friend class LogContext;
+  ValueBuilder(Vals vals) : _vals(std::move(vals)) {}
+  Vals _vals;
+};
+
+struct LogContext::Values {
+  virtual ~Values() = default;
+  virtual void visit(Visitor const&) const = 0;
+};
+
+template <class Vals, const char... Keys[]>
+struct LogContext::ValuesImpl final : Values {
+  explicit ValuesImpl(Vals vals) : _values(std::move(vals)) {}
+  void visit(Visitor const& visitor) const override {
+    doVisit<0, Keys...>(visitor);
+  }
+  ValuesImpl const* operator->() const { return this; }
+
+ private:
+  template <std::size_t Idx, const char K[], const char... Ks[]>
+  void doVisit(Visitor const& visitor) const {
+    visitor.visit(K, toVisitorType(std::get<Idx>(_values)));
+    if constexpr (Idx + 1 < std::tuple_size<Vals>{}()) {
+      doVisit<Idx + 1, Ks...>(visitor);
+    }
+  }
+  
+  template <class T>
+  static auto toVisitorType(T&& v) {
+    using TT = std::remove_reference_t<T>;
+    if constexpr (std::is_floating_point_v<TT>) {
+      return static_cast<double>(v);
+    } else if constexpr (std::is_unsigned_v<TT>) {
+      return static_cast<std::uint64_t>(v);
+    } else if constexpr (std::is_signed_v<TT>) {
+      return static_cast<std::int64_t>(v);
+    } else if constexpr (std::is_constructible_v<std::string_view, T>) {
+      return std::string_view(std::forward<T>(v));
+    } else if constexpr (std::is_constructible_v<std::string, T>) {
+      return std::string(std::forward<T>(v));
+    } else {
+      std::ostringstream ss;
+      ss << v;
+      return std::move(ss).str();
+    }
+  }
+
+  Vals _values;
+};
+
+struct LogContext::Entry {
+  virtual ~Entry() {
+    TRI_ASSERT(_refCount.load() <= 1);
+  }
+  virtual void visit(Visitor const&) const = 0;
+  
+ protected:
+  virtual void release(EntryCache& cache) noexcept {
+    this->~Entry();
+    ::operator delete(this);
+  }
+
+ private:
+  void incRefCnt() noexcept {
+    TRI_ASSERT(_refCount.load() > 0);
+    _refCount.fetch_add(1, std::memory_order_acquire);
```

Comment:
Instead of two separate load and add operations, we could fuse it to use a single add operation.
For release mode, this won't make a difference though, so feel free to reject the suggestion.
```suggestion
    [[maybe_unused]] auto old = _refCount.fetch_add(1, std::memory_order_acquire);
    TRI_ASSERT(old > 0);
```

Reply:
I prefer to have the assertion before the fetch_add so that the variable remains unchanged in case the assertion fails.